Add governance guardrails tests for update pparams

nabinpkl opened the pull request:
This PR proposes tests for governance guardrails. This is based on the Plutus Governance guardrails script and default constitution. At first, the script is enacted and then checked for unconstitutional values/out-of-range values for all protocol parameters defined in the constitution along with a check to ensure that valid values are accepted as a protocol parameter governance action.

The parameters and their values are checked based on Plutus cardano-constitution.

commit [db119872c0dfc8537a39d38cea8ef0b20c59ecb5](https://github.com/IntersectMBO/plutus/tree/db119872c0dfc8537a39d38cea8ef0b20c59ecb5/cardano-constitution)

The default constitution file
https://github.com/IntersectMBO/plutus/blob/db119872c0dfc8537a39d38cea8ef0b20c59ecb5/cardano-constitution/data/defaultConstitution.json

The user stories are derived from this PR.
https://github.com/IntersectMBO/cardano-test-plans/pull/52

This PR is ready for review and waiting for user stories to be merged.

mkoura reviewed (changes requested):
Thanks for the PR. So far I took only brief look and have several comments.

There can be improvements made to enable parallel run of the tests (they don't change the environment, apart from the setup, so I believe they can run in parallel) and collateral handling.

If you don't mind, I'll do the changes, it will be faster for me than writing detailed comments.

nabinpkl commented on cardano_node_tests/tests/common.py:
```diff
@@ -322,3 +322,38 @@ def get_conway_address_deposit(cluster_obj: clusterlib.ClusterLib) -> int:
     return stake_deposit_amt
+
+
+def create_collaterals(
```
Thank you for the feedback. This has been moved to clustelib_utils with the slicing of utxos on the latest update.

nabinpkl commented on cardano_node_tests/tests/common.py:
```diff
@@ -322,3 +322,38 @@ def get_conway_address_deposit(cluster_obj: clusterlib.ClusterLib) -> int:
+    return [
+        clusterlib.filter_utxos(utxos=utxos, utxo_ix=utxo_ix_offset + i)
+        for i in range(len(tx_outs))
+    ]
```
This has been moved to clustelib_utils with the slicing of utxos on the latest update.

nabinpkl commented on cardano_node_tests/tests/tests_conway/test_guardrails.py:
```diff
@@ -0,0 +1,1543 @@
+    )
+
+
+class ClusterWithConstitutionRecord:
```
Right. This has been updated to simpler dataclass.

nabinpkl commented on cardano_node_tests/tests/tests_conway/test_guardrails.py:
```diff
@@ -0,0 +1,1543 @@
+            action_type=governance_utils.PrevGovActionIds.CONSTITUTION,
+            gov_state=cluster.g_conway_governance.query.gov_state(),
+        )
+        constitution_action = cluster.g_conway_governance.action.create_constitution(
```
`propose_change_constitution` has been updated with additional optional `constitution_script_hash` parameter.

nabinpkl commented on cardano_node_tests/tests/tests_conway/test_guardrails.py:
```diff
@@ -0,0 +1,1543 @@
+        deposit_return_stake_vkey_file=pool_user_lg.stake.vkey_file,
+    )
+
+    # Create collateral utxo for plutus script
```
The collateral creation step has been moved to set up.

nabinpkl commented on cardano_node_tests/tests/tests_conway/test_guardrails.py:
```diff
@@ -0,0 +1,1543 @@
+    cluster_with_constitution: ClusterWithConstitutionRecord,
+    param: GuardrailTestParam,
+    min_value: int | float,
+    dependent_proposals: List[clusterlib_utils.UpdateProposal] | Tuple,
```
Right. It has been updated to use Union instead.

nabinpkl commented on cardano_node_tests/tests/tests_conway/test_guardrails.py:
```diff
@@ -0,0 +1,1543 @@
+        _enact_script_constitution()
+
+    # Testnet will be using an unexpected constitution, respin is needed
+    cluster_manager.set_needs_respin()
```
I am not familiar much with the framework you mentioned. This step has been removed and I believe it needs to be updated in the CI/CD pipeline for running guardrails test.

nabinpkl commented on cardano_node_tests/tests/tests_conway/test_guardrails.py:
```diff
@@ -0,0 +1,1543 @@
+):
+    """Check that the guardrails are enforced."""
+    with pytest.raises(clusterlib.CLIError) as excinfo:
+        propose_param_changes(cluster_with_constitution, proposals)
```
All the function calls with more than one argument are updated to use kwargs.

nabinpkl commented:
> Thanks for the PR. So far I took only brief look and have several comments.
> 
> There can be improvements made to enable parallel run of the tests (they don't change the environment, apart from the setup, so I believe they can run in parallel) and collateral handling.
> 
> If you don't mind, I'll do the changes, it will be faster for me than writing detailed comments.

Thank you for the review comments and feedback.
I have addressed some of the comments.

We'd welcome the changes you'd like to make to the PR.

It'd be great if you could push the changes directly on the PR, get it merged if possible, or merge PR and then make the necessary changes to run them in parallel. 

Or I can address those changes If there are some examples of tests currently running in parallel and how the setup is being done.

mkoura commented:
Looks like I cannot push to your PR. Can you give me access?

nabinpkl commented:
> Looks like I cannot push to your PR. Can you give me access?

I just sent you an invite. It should give access now.
